### DISCUSSION TITLE
Sentry into Microfrontend

### DISCUSSION THREAD

**Discussion by seema-0611** [web-infra-dev/rspack#8353] [General]
How do we integrate Sentry into a microfrontend project built with React and Rspack as the bundler?

**Comment by markmssd** [comment]
Have you tried https://docs.sentry.io/platforms/javascript/best-practices/micro-frontends?

**Reply by seema-0611** [reply]
There is no specific documentation available for Rspack as a bundler.

**Reply by markmssd** [reply]
According to https://rspack.dev/guide/compatibility/plugin, the Sentry webpack plugin should be compatible.

**Reply by markmssd** [reply]
Worth noting that there was issues with it though: https://github.com/web-infra-dev/rspack/issues/7914, but it should be fixed?

**Reply by seema-0611** [reply]
Could you please elaborate this method in deep - https://docs.sentry.io/platforms/javascript/best-practices/micro-frontends ?
Could you refer some example of this implementation?

**Reply by seema-0611** [reply]
I tried to implement this, but it didn't work.
Errors from different microfrontend repo and shell is logged in same Sentry project only.

**Reply by markmssd** [reply]
Could you provide example configs of what you've tried, and what errors you're getting? A bit hard to help otherwise. Also, if your error is not Rspack specific, you might have better chance asking in Sentry's [repo](https://github.com/getsentry/sentry-javascript-bundler-plugins) directly :)

**Reply by seema-0611** [reply]
I am not sure , if I am doing it in correct way.
In the shell repo , I have initialized Sentry like this.

import React from 'react';
import ReactDOM from 'react-dom/client';
import { BrowserRouter } from 'react-router-dom';
import App from './App';
import { BaseLayout } from './components';
import { GlobalProvider } from './context';
import ErrorBoundary from './ErrorBoundary';
import './index.css';
import {
  init,
  makeFetchTransport,
  moduleMetadataIntegration,
  makeMultiplexedTransport,
} from "@sentry/browser";

const EXTRA_KEY = "ROUTE_TO";

const transport = makeMultiplexedTransport(makeFetchTransport, (args) => {
  const event = args.getEvent();
  if (
    event &&
    event.extra &&
    EXTRA_KEY in event.extra &&
    Array.isArray(event.extra[EXTRA_KEY])
  ) {
    return event.extra[EXTRA_KEY];
  }
  return [];
});

init({
  dsn: "_DEFAULT_DSN",
  environment: process.env.NODE_ENV,
  integrations: [moduleMetadataIntegration()],
  transport,
  beforeSend: (event) => {
    if (event?.exception?.values?.[0].stacktrace.frames) {
      const frames = event.exception.values[0].stacktrace.frames;
      // Find the last frame with module metadata containing a DSN
      const routeTo = frames
        .filter(
          (frame) => frame.module_metadata && frame.module_metadata.dsn,
        )
        .map((v) => v.module_metadata)
        .slice(-1); // using top frame only - you may want to customize this according to your needs

      if (routeTo.length) {
        event.extra = {
          ...event.extra,
          [EXTRA_KEY]: routeTo,
        };
      }
    }

    return event;
  },
});

const rootEl = document.getElementById('root');
if (rootEl) {
  const root = ReactDOM.createRoot(rootEl);
  root.render(
    <React.StrictMode>
      <GlobalProvider>
        <BrowserRouter basename={process.env.PUBLIC_BASE_PATH}>
          <ErrorBoundary>
            <BaseLayout>
              <App />
            </BaseLayout>
          </ErrorBoundary>
        </BrowserRouter>
      </GlobalProvider>
    </React.StrictMode>
  );
}
In webpack.config.ts
const { sentryWebpackPlugin } = require("@sentry/webpack-plugin");

module.exports = {
  // ... other options
  devtool: "source-map", // Source map generation must be turned on
  plugins: [
    // Put the Sentry Webpack plugin after all other plugins
    sentryWebpackPlugin({
      authToken: process.env.SENTRY_AUTH_TOKEN,
      org: "org",
      project: "project,
      moduleMetadata: ({ release }) => ({
          dsn: "_MODULE_DSN",
          release,
        }),
     
    }),
  ],
and in another microfrontend project , I have only add webpack.config file with different DSN.

The issue is that if I am throwing error in my other microfrontend project , It is captured by  DSN of shell project.It is not differentiating among the microfronted.

**Reply by markmssd** [reply]
You can try to `console.log(routeTo)` in the frontend, and see if it correctly has the `release/dsn` from the other microfrontend.

**Reply by markmssd** [reply]
Also, make sure that you have selected the `Latest` issue in the dropdown, it might help seeing what you're expecting:

<img width="1428" alt="Screenshot 2024-11-07 at 12 05 46 PM" src="https://github.com/user-attachments/assets/7944c8e3-320a-4ede-91bb-9d70b089a38f">

**Reply by seema-0611** [reply]
I have checked it , stack trace frame do not have module_metadata property in it. I guess metadata is not correctly injected.
As a result , toroute is an empty array.

**Reply by markmssd** [reply]
Make sure you're on the latest version of the plugin, and if it's still not working, you should open an issue in their repo!
